pkg/apis/pipeline/v1/task_validation.go

```diff
@@ -362,39 +362,69 @@ func (p ParamSpec) ValidateObjectType(ctx context.Context) *apis.FieldError {
 	return nil
 }
 
-// ValidateParameterVariables validates all variables within a slice of ParamSpecs against a slice of Steps
-func ValidateParameterVariables(ctx context.Context, steps []Step, params []ParamSpec) *apis.FieldError {
-	allParameterNames := sets.NewString()
-	stringParameterNames := sets.NewString()
-	arrayParameterNames := sets.NewString()
-	objectParamSpecs := []ParamSpec{}
-	var errs *apis.FieldError
+// sortParamsByType splits the input params into string params, array params, and object params, in that order
+func sortParamsByType(params []ParamSpec) ([]ParamSpec, []ParamSpec, []ParamSpec) {
+	var stringParams, arrayParams, objectParams []ParamSpec
 	for _, p := range params {
-		// validate no duplicate names
-		if allParameterNames.Has(p.Name) {
-			errs = errs.Also(apis.ErrGeneric("parameter appears more than once", "").ViaFieldKey("params", p.Name))
-		}
-		allParameterNames.Insert(p.Name)
-
 		switch p.Type {
 		case ParamTypeArray:
-			arrayParameterNames.Insert(p.Name)
+			arrayParams = append(arrayParams, p)
 		case ParamTypeObject:
-			objectParamSpecs = append(objectParamSpecs, p)
+			objectParams = append(objectParams, p)
 		case ParamTypeString:
 			fallthrough
 		default:
-			stringParameterNames.Insert(p.Name)
+			stringParams = append(stringParams, p)
 		}
 	}
-	errs = errs.Also(validateNameFormat(stringParameterNames.Insert(arrayParameterNames.List()...), objectParamSpecs))
+	return stringParams, arrayParams, objectParams
+}
+
+func getNames(params []ParamSpec) []string {
+	var names []string
+	for _, p := range params {
+		names = append(names, p.Name)
+	}
+	return names
+}
```

Comment:
can this be a member function of `ParamSpecs`? I'd imagine that this is useful elsewhere, and most likely this code  is duplicated

Reply:
done